torchok/models/heads/segmentation/hrnet_head.py

```diff
@@ -22,19 +147,97 @@ def __init__(self, in_channels: int, num_classes: int):
+        x = self.convbnact(feats)
+        out = self.final_conv_layer(x)
+        out = resize(out, input_image.shape[2:])
+        return out
+
+
+@HEADS.register_class
+class HRNetOCRSegmentationHead(BaseModel):
+    has_ocr = True
+
+    def __init__(self, in_channels: int, num_classes: int, ocr_mid_channels=128,
+                 ocr_key_channels=64, **kwargs):
+        super(HRNetOCRSegmentationHead, self).__init__(in_channels, num_classes)
+        self.num_classes = num_classes
+
+        self.conv3x3_ocr = ConvBnRelu(in_channels, ocr_mid_channels, kernel_size=3, padding=1)
+        self.ocr_gather_head = SpatialGather_Module(num_classes)
+
+        self.ocr_distri_head = SpatialOCR(in_channels=ocr_mid_channels,
+                                          key_channels=ocr_key_channels,
+                                          out_channels=ocr_mid_channels,
+                                          scale=1,
+                                          dropout=0.05,
+                                          )
+        self.last_reduction = ConvBnRelu(
+            ocr_mid_channels,
+            ocr_mid_channels // 16,
+            kernel_size=1, stride=1, padding=0
+        )
+
+        self.cls_head = nn.Conv2d(ocr_mid_channels // 16, num_classes,
+                                  kernel_size=1, stride=1, padding=0, bias=True)
+
+        self.aux_head = nn.Sequential(
+            ConvBnRelu(in_channels, in_channels,
+                       kernel_size=1, stride=1, padding=0),
+            nn.Conv2d(in_channels, num_classes,
+                      kernel_size=1, stride=1, padding=0, bias=True)
+        )
+
+        self.init_weights()
+
+    def init_weights(self):
+        for name, m in self.named_modules():
+            if isinstance(m, nn.Conv2d):
+                nn.init.kaiming_uniform_(m.weight, mode="fan_in", nonlinearity="relu")
+                if m.bias is not None:
+                    nn.init.constant_(m.bias, 0)
+            elif isinstance(m, nn.BatchNorm2d):
+                nn.init.constant_(m.weight, 1)
+                nn.init.constant_(m.bias, 0)
+            elif isinstance(m, nn.Linear):
+                nn.init.xavier_uniform_(m.weight)
+                if m.bias is not None:
+                    nn.init.constant_(m.bias, 0)
+
+    def forward(self, feats):
+        input_image, feats = feats
+
+        # ocr
+        out_aux = self.aux_head(feats)
+        # compute contrast feature
+        feats = self.conv3x3_ocr(feats)
+
+        context = self.ocr_gather_head(feats, out_aux)
+        feats = self.ocr_distri_head(feats, context)
+
+        feats = self.last_reduction(feats)
+        out = self.cls_head(feats)
+
+        out = resize(out, input_image.shape[2:])
+        out_aux = resize(out_aux, input_image.shape[2:])
+
+        if self.num_classes == 1:
+            out = out[:, 0]
+            out_aux = out_aux[:, 0]
+
+        if self.training:
+            return out, out_aux
```

Comment:
Why we need out_aux in training?

Reply:
Because it is how OCR is trained correctly